### PR TITLE
Draw Operational area  boundaries

### PR DESCRIPTION
**Pull request by githengi**
Fixes 
- [https://github.com/OpenSRP/opensrp-client-reveal/issues/128](https://github.com/OpenSRP/opensrp-client-reveal/issues/128)
- [https://github.com/OpenSRP/opensrp-client-reveal/issues/124](https://github.com/OpenSRP/opensrp-client-reveal/issues/124)

### CODE REVIEW

File: opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java [hunk 1]
```diff
@@ -365,19 +367,26 @@ public void setGeoJsonSource(@NonNull FeatureCollection featureCollection, Featu
...
                 }
-                if (boundaryBuilder == null) {
-                    boundaryBuilder = new BoundaryLayer.Builder(FeatureCollection.fromFeature(operationalArea))
-                            .setLabelProperty(Map.NAME_PROPERTY)
-                            .setLabelTextSize(getResources().getDimension(R.dimen.operational_area_boundary_text_size))
-                            .setLabelColorInt(Color.WHITE)
-                            .setBoundaryColor(Color.WHITE)
-                            .setBoundaryWidth(getResources().getDimension(R.dimen.operational_area_boundary_width));
+                if (boundaryLayer == null) {
+                    boundaryLayer = createBoundaryLayer(operationalArea);
```
Comment (vincent-karuri) [inline opensrp-reveal/src/main/java/org/smartregister/reveal/view/ListTasksActivity.java]: Move this outside the if block

**Comment by vincent-karuri** [comment]
Are there suppose to be two labels for the operational area name?

**Comment by vincent-karuri** [comment]
I'm unable to add a new structure from the form.

**Comment by vincent-karuri** [comment]
```
04-08 10:15:06.852 9495-9753/org.smartregister.reveal E/AndroidRuntime: FATAL EXCEPTION: pool-5-thread-1
    Process: org.smartregister.reveal, PID: 9495
    java.lang.NullPointerException: Attempt to invoke virtual method 'com.google.gson.JsonObject com.mapbox.geojson.Feature.properties()' on a null object reference
        at com.mapbox.geojson.Feature.fromJson(Feature.java:74)
        at org.smartregister.reveal.interactor.ListTaskInteractor$3.run(ListTaskInteractor.java:211)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
        at java.lang.Thread.run(Thread.java:818)
04-08 10:15:06.982 1721-9791/? E/android.os.Debug: ro.product_ship = true
    ro.debug_level = 0x4f4c
    sys.mobilecare.preload = false
04-08 10:15:07.982 9495-9495/org.smartregister.reveal E/WindowManager: android.view.WindowLeaked: Activity org.smartregister.reveal.view.ListTasksActivity has leaked window com.android.internal.policy.PhoneWindow$DecorView{64cfa77 V.E...... R......D 0,0-668,347} that was originally added here
        at android.view.ViewRootImpl.<init>(ViewRootImpl.java:603)
        at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:326)
        at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:109)
        at android.app.Dialog.show(Dialog.java:505)
        at org.smartregister.reveal.view.ListTasksActivity.showProgressDialog(ListTasksActivity.java:522)
        at org.smartregister.reveal.presenter.ListTaskPresenter.onDrawerClosed(ListTaskPresenter.java:275)
        at org.smartregister.reveal.view.ListTasksActivity$4.onDrawerClosed(ListTasksActivity.java:240)
        at android.support.v4.widget.DrawerLayout.dispatchOnDrawerClosed(DrawerLayout.java:800)
        at android.support.v4.widget.DrawerLayout.updateDrawerState(DrawerLayout.java:770)
        at android.support.v4.widget.DrawerLayout$ViewDragCallback.onViewDragStateChanged(DrawerLayout.java:2077)
        at android.support.v4.widget.ViewDragHelper.setDragState(ViewDragHelper.java:887)
        at android.support.v4.widget.ViewDragHelper$2.run(ViewDragHelper.java:341)
        at android.os.Handler.handleCallback(Handler.java:739)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:148)
        at android.app.ActivityThread.main(ActivityThread.java:7325)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
04-08 10:15:08.152 1721-2613/? E/Watchdog: !@Sync 51 [04-08 10:15:08.160]
04-08 10:15:28.622 9495-9764/org.smartregister.reveal E/org.smartregister.sync.helper.TaskServiceHelper: Invalid long: ""
    java.lang.NumberFormatException: Invalid long: ""
        at java.lang.Long.invalidLong(Long.java:124)
        at java.lang.Long.parseLong(Long.java:345)
        at java.lang.Long.parseLong(Long.java:321)
        at org.smartregister.sync.helper.TaskServiceHelper.fetchTasksFromServer(TaskServiceHelper.java:73)
        at org.smartregister.sync.helper.TaskServiceHelper.syncTasks(TaskServiceHelper.java:65)
        at org.smartregister.reveal.sync.LocationTaskIntentService.doSync(LocationTaskIntentService.java:47)
        at org.smartregister.reveal.sync.LocationTaskIntentService.onHandleIntent(LocationTaskIntentService.java:38)
        at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:66)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:148)
        at android.os.HandlerThread.run(HandlerThread.java:61) 
```

Got this runtime exception after selecting an operational area after a fresh install.

**Comment by githengi** [comment]
> Are there suppose to be two labels for the operational area name?

This is because on the raster  satellite tiles provided have name and boundaries already drawn.

Test with Cambodia OA which does not use custom satellite tiles and use only mapbox base map.  

I created an issue for optionally disabling drawing operational area and boundary if they are on the base map tiles 

[https://github.com/OpenSRP/opensrp-client-reveal/issues/130](https://github.com/OpenSRP/opensrp-client-reveal/issues/130)

**Comment by githengi** [comment]
> I'm unable to add a new structure from the form.

Share an error trace or mockup, I am not able to reproduce this

**Comment by vincent-karuri** [comment]
> > I'm unable to add a new structure from the form.
> 
> Share an error trace or screenshot, I am not able to reproduce this

It works now, it may have had something to do with the intermittent internet and location services 🤷🏾‍♂️

**Comment by vincent-karuri** [comment]
Screenshot_20190408-151927

This issue seems to be intermittent and I didn't see any relevant logs.

**Comment by vincent-karuri** [comment]
image

This was the entry that got rejected

**Comment by vincent-karuri** [comment]
It seems to happen when one tries to add a point at the wrong zoom level.

The first time it shows a toast message but after that, I cannot add a point even after zooming in.

**Comment by githengi** [comment]
The issue has been resolved, Geowidget is now being validate even if form in invalid due to native form refactor

File: opensrp-reveal/src/main/java/org/smartregister/reveal/presenter/RevealJsonFormFragmentPresenter.java [hunk 2]
```diff
@@ -51,26 +51,25 @@ public RevealJsonFormFragmentPresenter(JsonFormFragment formFragment, JsonFormIn
...
+                    getInvalidFields().put(fieldKey, validationStatus);
+                } else {
+                    getInvalidFields().remove(fieldKey);
+                    if (isFormValid() && validateFarStructures()) {
+                        validateUserLocation(mapView);
+                    } else if (isFormValid()) {
+                        onLocationValidated();
                     }
-                    break;//exit loop, assumption; there will be only 1 map per form.
+                    return;
```
Comment (vincent-karuri) [inline opensrp-reveal/src/main/java/org/smartregister/reveal/presenter/RevealJsonFormFragmentPresenter.java]: Why is this needed if there is a break anyway?

**Comment by vincent-karuri** [comment]
Try adding a point, the password dialog appears and immediately disappears without the password being entered.

The point is however added.
